catalog/resource_connection.go

```diff
@@ -0,0 +1,112 @@
+package catalog
+
+import (
+	"context"
+
+	"github.com/databricks/databricks-sdk-go/service/catalog"
+	"github.com/databricks/terraform-provider-databricks/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"golang.org/x/exp/slices"
+)
+
+// This structure contains the fields of catalog.UpdateConnection and catalog.CreateConnection
+// We need to create this because we need Owner, FullNameArg, SchemaName and CatalogName which aren't present in a single of them.
+// We also need to annotate tf:"computed" for the Owner field.
+type ConnectionInfo struct {
+	// User-provided free-form text description.
+	Comment string `json:"comment,omitempty" tf:"force_new"`
+	// The type of connection.
+	ConnectionType string `json:"connection_type" tf:"force_new"`
+	// Name of the connection.
+	Name string `json:"name"`
+	// Name of the connection.
+	NameArg string `json:"-" url:"-"`
+	// A map of key-value properties attached to the securable.
+	Options map[string]string `json:"options" tf:"sensitive"`
+	// Username of current owner of the connection.
+	Owner string `json:"owner,omitempty" tf:"force_new,suppress_diff"`
+	// An object containing map of key-value properties attached to the
+	// connection.
+	Properties map[string]string `json:"properties,omitempty" tf:"force_new"`
+	// If the connection is read only.
+	ReadOnly bool `json:"read_only,omitempty" tf:"force_new,computed"`
+}
+
+// suppress diff for sensitive options, which are not returned by the server
+func suppressSensitiveOptions(k, old, new string, d *schema.ResourceData) bool {
+	//this list will expand as other auth may have different sensitive options
+	sensitiveOptions := []string{"user", "password"}
```

Comment:
Yes, but the question is how to determine whether user/password options are different locally than what is configured (imagine someone changed user/password manually in Databricks or via a separate process). How can we compute whether or not the resource needs to be updated? Typically, TF reads resources to compare them to its state & the configuration supplied by the user to compute a diff. If `user` and `password` are not returned, it may not be possible to know whether the connection needs to be updated.

Reply:
Is it possible to always update the username and password? There are basically 2 worlds:
1. Silently fail to update when user changes username/password
2. Always update username / password.

Reply:
I am inclined not to accept either option. The former would lead to an exceptionally poor user experience, where they would need to do manual changes to rotate a credential. The latter is also not great, because it will indicate that there is always a diff even when there is no change needed to the resource, and in general will block users from using this resource if they have checks to ensure that there are no diffs between configured and deployed state.

My preference would be to override the CustomizeDiff function to: 1) diff all other fields, and then 2) check whether options that are considered secrets are different between config and state via `(r *schema.ResourceDiff) GetChange(k string)` method.

We do need to figure out a consistent story for dealing with secrets. In workspaces, we have secret management. Have we thought about how this would work for UC at the account-level?

Reply:
@mgyucht 

>My preference would be to override the CustomizeDiff function to: 1) diff all other fields, and then 2) check whether options that are considered secrets are different between config and state via (r *schema.ResourceDiff) GetChange(k string) method.

Sorry I am a bit ignorant - how does `(r *schema.ResourceDiff) GetChange(k string)` method decide there's a diff? Via some hash object?

Reply:
Terraform maintains a state file with all of the attributes from each resource it creates. The Terraform lifecycle is divided into two parts roughly speaking:
1. Figure out what resources need to be created, updated and/or deleted based on the differences between the configuration, the TF state file, and the state of the remote resource;
2. Apply those changes.
`GetChange()` by my reading returns, for a given field in the TF schema for the resource, the "old" value from the state and the "new" value from the config + remote state. I have to think through how this works a bit carefully, but we might be able to do something like: check if the config specifically has differences from the state for secret option keys, and for other keys, just check if there is a difference between the config and the remote state.

Reply:
I see, so terraform remembers the state of last time that customer creates, makes sense.

Reply:
Yeah then I would say maybe let's just go with the simple approach and just always refresh.

Reply:
Given that we have the choice between forcing users to make manual changes if secret fields are updated vs updating them every time, I agree that we should lean towards always refreshing. In light of moving forward with this resource and unblocking customers, let's do this. @nkvuong what do you think, would this be a viable option from the Terraform perspective? IIUC to do this, we treat the secret options in the same way as non-secret options, and they will be included in the diff every time. 

@andrewli81 I really recommend that your team revisits the secret management story for UC connections. We don't need each team to reinvent the wheel when it comes to secret management, and when this happens, it increases complexity for consumers and results in worse user experiences. For inspiration, you can look at how [secrets are referenced in Spark configurations](https://docs.databricks.com/en/security/secrets/secrets.html#use-a-secret-in-a-spark-configuration-property-or-environment-variable). I understand that there is a workspace-level/account-level divide that will need to be addressed. There may be other use-cases for account-level secrets in the future as well (or even today: consider account-level service principal passwords).

Reply:
> we treat the secret options in the same way as non-secret options, and they will be included in the diff every time.

this will cause permanent diff in the Terraform apply - we will need to add to doc that users should add `ignore_changes` to their resource definition to avoid that.

Reply:
we have similar problem with MLflow webhooks, and on read we simply restore data from the state: https://github.com/databricks/terraform-provider-databricks/blob/master/mlflow/resource_mlflow_webhook.go#L79 - we can try to do the same here

Reply:
>I understand that there is a workspace-level/account-level divide that will need to be addressed. There may be other use-cases for account-level secrets in the future as well (or even today: consider account-level service principal passwords).

We have to be unblocked on shipping features that need to store secrets in UC account level objects in lieu of a team that is willing to fund an account level secret manager, I will probably write a vision doc for this because I think we will need it for other use cases as well.